Lazy load post meta

spacedmonkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket:  https://core.trac.wordpress.org/ticket/57496

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Lazy load post meta

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Rebase

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Fix

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Update callback name and implement lazyload_post_meta function

Renamed 'lazyload_meta_callback' to 'lazyload_post_meta' in 'post' array. Introduced new function 'lazyload_post_meta' for lazy-loading metadata of queued posts. The function also includes logic for resetting the object queue after the meta cache has been updated.

spacedmonkey committed:
Update expected database queries in tests

The commit updates the expected number of database queries in several PHPUnit tests. The changes reflect improvements made in database query management and caching strategy. As a result, these tests are now expecting fewer queries than before, ensuring that our performance optimizations are correctly implemented and functioning.

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Add filter for REST API menu read access check

Introduce 'rest_menu_read_access' filter to allow customization of read access permissions in REST API endpoints for menus, menu items, and menu locations. This facilitates more flexible access control based on custom criteria.

spacedmonkey committed:
Update post meta cache function call

Replaced wp_lazyload_post_meta with update_postmeta_cache to improve clarity and ensure the correct function is executed for updating post meta cache.

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Revert these changes.

spacedmonkey commented on tests/phpunit/tests/post/nav-menu.php:
```diff
@@ -342,7 +342,7 @@ public function test_wp_get_nav_menu_items_cache_primes_posts() {
 		wp_get_nav_menu_items( $this->menu_id, array( 'nopaging' => false ) );
 		$queries_made = get_num_queries() - $start_num_queries;
-		$this->assertSame( 7, $queries_made, 'Only does 7 database queries when running wp_get_nav_menu_items.' );
+		$this->assertSame( 6, $queries_made, 'Only does 7 database queries when running wp_get_nav_menu_items.' );
```
```suggestion
		$this->assertSame( 6, $queries_made, 'Only does 6 database queries when running wp_get_nav_menu_items.' );
```

spacedmonkey committed:
Apply suggestions from code review

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Add $lazy_load_post_meta parameter to WP_Query and use it in post controller.

spacedmonkey committed:
Add $lazy_load_post_meta parameter in wp_get_post_revisions.

spacedmonkey committed:
Fix tests.

spacedmonkey committed:
Improve tests.

spacedmonkey committed:
Fix tests again.

spacedmonkey committed:
Add tests.

spacedmonkey committed:
wp_dashboard_recent_posts lazy load post meta.

joemcgill commented on src/wp-includes/class-wp-metadata-lazyloader.php:
```diff
@@ -163,6 +167,41 @@ public function lazyload_comment_meta( $check ) {
+
+		$object_ids = array_keys( $this->pending_objects[ $meta_type ] );
+		if ( $object_id && ! in_array( $object_id, $object_ids, true ) ) {
+			return $check;
```
This whole method is a line for line duplication of `lazyload_meta_callback()` except for this line. Why is the logic different here, and could this be simplified by adding a check for `$meta_type` to return the `$check` rather than adding the `$object_id` to the array of IDs?

spacedmonkey replied:
@joemcgill This is the secret sause of this change. Unlike other meta types like, site, comment and term, post meta gets called ALOT. So to stop every call to post meta loading this lazy load queue, this only loads if the queue if an item is in the post meta queue. 

For, if I have 10 items of post meta loaded and 5 items in the lazy load queue, if you request an post id in the load meta, you DONT want the queue items to load.

peterwilsoncc replied:
@spacedmonkey I think the logic here should be slightly different. If the object ID passed to `get_post_meta()` is already cached, then it makes sense not to load the queued data. If the object ID is not cached, then a database call will be made and it makes sense to load the queued data.

spacedmonkey replied:
@peterwilsoncc that is not how the other meta types work. IDs are put in the queue. If the id is already loaded, when wp_cache_get_multiple, is called, if it is only object is not cache is loaded. 

I don't see lots of cases where an id in the queue but also primed. I expect that ids that might be used will be added to the queue, for example attachment meta or post parent meta.

peterwilsoncc replied:
@spacedmonkey I realise that is not how it works for other meta data (potentially a bug), but you are proposing that post meta works differently. 

However, the code in it's current form will trigger a database query for `get_post_meta( $uncached_post_id )` without priming other data in the queue. As the idea of lazy loading meta data is to ride another database query, it seems we should do that here even if the uncached object is not queued.

spacedmonkey replied:
@peterwilsoncc So you are saying, we should not have special behavour for post meta, it should work like other meta? That every time `get_post_meta` is called it should clear the queue? If that is the case, as post meta is called so much, then there is no performance benefit. The queue would be cleared so much, that we might as well not bother. 

It is less about saving database queries I guess. It useful in the following use case. 

I have loaded 5 posts and there meta. But I have also loaded, all those post parents, ( say for revisions or attachments ). Also add the parents post meta to the queue, as it MAY get used. In cases, where it is, load them all at once, in other cases the items in the queue may never get loaded, saving a database query. There are LOTs of places in core, where post meta is loaded, because it MAY be used a plugin, ( example if you have a filter of the title to change it a value loaded for post meta ). But core itselve is not using this post meta. Once we have a queue, where it says maybe load post meta, we can stop priming post meta in lots of places where it is not required by core, but turning off the priming might effect sites with plugins that except the post meta to be loaded.

peterwilsoncc replied:
> That every time `get_post_meta` is called it should clear the queue? If that is the case, as post meta is called so much, then there is no performance benefit. The queue would be cleared so much, that we might as well not bother.

No, this is not what I am saying.

What I am saying is that IF a database query is required then the queue should be cleared. For example:

Cached post meta IDs: 1, 2, 3, 4
Queued post meta IDs: 5, 6, 7, 8

`get_post_meta( 1 )` would not clear the queue as it does not require a database call
`get_post_meta( 10 )` does require a database call so would use the opportunity to clear the queue and prime the cache for the queued meta data.

spacedmonkey replied:
Okay, I think I am starting to get that. I am honestly not sure how we would do this. 

I also think there is a trade off here between number of database queries vs memory usage. I see the desire to limit the number of database queries here. But loading post meta that might not ever be used could be worse than before. The current patch ensure that the post meta is used, as it is requested.

peterwilsoncc replied:
I agree, memory use is a risk and possibly why the lazy loading of post meta wasn't introduced earlier but I haven't been able to find an explanation.

Using my example above, the risk exists for queued meta `get_post_meta( 8 )` too, so I am not sure that it's a problem to accept the risk for the unqueued meta that requires a database call `get_post_meta( 10 );`

spacedmonkey replied:
@peterwilsoncc I am not sure I understand what you are trying to say here. What is your recommended course of action?

peterwilsoncc replied:
@spacedmonkey I am saying that if a database query is to be made, the queue should be processed. This should happen regardless of whether the object ID is in the queue or not.

joemcgill replied:
Thanks for addressing my suggestion about consolidating this logic into the existing method in [ebf588e](https://github.com/WordPress/wordpress-develop/pull/4216/commits/ebf588e3f8a814672abac397ea2970309583b1d3). Once the approach for clearing the queue has been settled, I'm happy with the shape of this update.

joemcgill commented on src/wp-includes/class-wp-query.php:
```diff
@@ -1967,6 +1972,12 @@ public function get_posts() {
 		}
 
+		if ( ! isset( $q['lazy_load_post_meta'] ) ) {
+			$q['lazy_load_post_meta'] = false;
```
Per the DocBlock, shouldn't this be the following?

```suggestion
			$q['lazy_load_post_meta'] = $q['update_post_meta_cache'];
```

joemcgill replied:
Rather than just removing the docs about the default in [4e3a8ec](https://github.com/WordPress/wordpress-develop/pull/4216/commits/4e3a8ecbb8a172501cefe184ca413d929fe4239a), can it be updated to be accurate?

spacedmonkey replied:
@joemcgill I am not sure what is being asked here.

joemcgill commented on src/wp-includes/class-wp-query.php:
```diff
@@ -1967,6 +1972,12 @@ public function get_posts() {
+		if ( ! isset( $q['lazy_load_post_meta'] ) ) {
+			$q['lazy_load_post_meta'] = false;
+		} elseif ( $q['lazy_load_post_meta'] ) {
+			$q['update_post_meta_cache'] = false;
```
For term meta, lazy loading requires term cache priming. Can you add an inline comment about why this is different? 

Also, the Docblock states...

> Setting to false will disable cache priming for post meta, so that each get_post_meta() call will hit the database.

...but that doesn't seem to be implemented.

joemcgill replied:
@spacedmonkey it still doesn't seem that this comment was addressed:

> Also, the Docblock states...
>
> > Setting to false will disable cache priming for post meta, so that each get_post_meta() call will hit the database.
>
> ...but that doesn't seem to be implemented.

Am I missing something?

spacedmonkey replied:
@joemcgill I have added some inline comments.

joemcgill reviewed (commented):
Thanks @spacedmonkey. I haven't tested this, but have left a few questions after reading through the proposed changes.

spacedmonkey commented on src/wp-includes/class-wp-query.php:
```diff
@@ -784,6 +785,10 @@ public function fill_query_vars( $query_vars ) {
+	 *     @type bool            $lazy_load_post_meta    Whether to lazy-load post meta. Setting to false will
+	 *                                                   disable cache priming for post meta, so that each
+	 *                                                   get_post_meta() call will hit the database.
+	 *                                                   Defaults to the value of `$update_post_meta_cache`.
```
```suggestion
```

spacedmonkey committed:
Apply suggestions from code review

spacedmonkey commented:
@joemcgill @mukeshpanchal27 mind taking a look at this PR?

spacedmonkey commented on src/wp-includes/post.php:
```diff
@@ -2616,6 +2616,21 @@ function get_post_meta( $post_id, $key = '', $single = false ) {
+/**
+ * Queue post meta for lazy-loading.
+ *
+ * @since 6.3.0
```
```suggestion
 * @since 6.8.0
```

spacedmonkey commented on src/wp-includes/class-wp-metadata-lazyloader.php:
```diff
@@ -163,6 +167,41 @@ public function lazyload_comment_meta( $check ) {
+	 * This method is public so that it can be used as a filter callback. As a rule, there
+	 * is no need to invoke it directly.
+	 *
+	 * @since x.x.x
```
```suggestion
	 * @since 6.8.0
```

spacedmonkey committed:
Apply suggestions from code review

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

joemcgill reviewed (commented):
I've taken another look at this and still seems like a few questions need to be addressed:

- https://github.com/WordPress/wordpress-develop/pull/4216#discussion_r1953459962
- https://github.com/WordPress/wordpress-develop/pull/4216#discussion_r1953464121

Also, seems like there is still some discussion that needs to be resolved in [this thread](https://github.com/WordPress/wordpress-develop/pull/4216#discussion_r1878485884) about when the queue should be cleared.

Even so, my original question in this thread didn't seem to be answered, which is whether `lazyload_meta_callback()` could be extended to handle lazy loading of post meta, since so much of the logic is duplicated. Something like this:

```diff
diff --git a/src/wp-includes/class-wp-metadata-lazyloader.php b/src/wp-includes/class-wp-metadata-lazyloader.php
index f9c02391d2..2d11e5f6ea 100644
--- a/src/wp-includes/class-wp-metadata-lazyloader.php
+++ b/src/wp-includes/class-wp-metadata-lazyloader.php
@@ -187,6 +187,11 @@ class WP_Metadata_Lazyloader {
 
 		$object_ids = array_keys( $this->pending_objects[ $meta_type ] );
 		if ( $object_id && ! in_array( $object_id, $object_ids, true ) ) {
+			// Only load in the post meta value is in the queue.
+			if ( 'post' === $meta_type ) {
+				return $check;
+			}
+
 			$object_ids[] = $object_id;
 		}
 
```

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-load-post-meta

spacedmonkey committed:
Refactor lazy-loading to unify metadata callback.

Replaces the `lazyload_post_meta` method with a consolidated `lazyload_meta_callback` for all meta types. Simplifies code structure and eliminates redundancy, while maintaining functionality for post metadata handling.

spacedmonkey committed:
Simplify post meta priming behavior and adjust formatting.

Clarifies the behavior of the `$lazy_load_post_meta` parameter by specifying fallback to default post meta priming. Additionally, adjusts whitespace in `tax_query` array assignments for consistency. These changes improve code readability and maintainability.

spacedmonkey committed:
Set default lazy loading for post meta to false

Defaulted `lazy_load_post_meta` to false and clarified its dependency on `update_post_meta_cache`. Ensured correct functionality by disabling post meta caching when lazy loading is enabled.

spacedmonkey committed:
Fix alignment in lazy load post meta test array.

Aligned the array key for better readability and consistency with the surrounding code. This change does not impact functionality but improves code clarity in the test file.

spacedmonkey reviewed (commented):
@joemcgill I believe I have actioned your feedback.

peterwilsoncc commented on src/wp-includes/class-wp-metadata-lazyloader.php:
```diff
@@ -187,6 +191,10 @@ public function lazyload_meta_callback( $check, $object_id, $meta_key, $single,
 
 		$object_ids = array_keys( $this->pending_objects[ $meta_type ] );
 		if ( $object_id && ! in_array( $object_id, $object_ids, true ) ) {
+			// Only load in the post meta value is in the queue.
+			if ( 'post' === $meta_type ) {
+				return $check;
+			}
```
@spacedmonkey Re https://github.com/WordPress/wordpress-develop/pull/4216#discussion_r1966428938, I think the code should simply be this to see if a database query is required. If not, it bypasses processing the queue, if so it adds the object ID and continues lazy loading.

```suggestion
			if ( false !== wp_cache_get( $object_id, $meta_type . '_meta' ) ) {
				// Object already in cache, no database query will be made.
				return $check;
			}
```

spacedmonkey replied:
@peterwilsoncc This happens in here 
https://github.com/WordPress/wordpress-develop/blob/bcb59bfb14cbfa79335fc9e411aae56b060da85d/src/wp-includes/meta.php#L1168-L1189

So if it first checks if the id is in cache, then if not, then it will do a database query. 

Add this check, just repeats to the logic.

peterwilsoncc replied:
It's not quite the same.

`get_metadata( type, **cached** object ID ...)` will process the queue, even though a database query is not required. This is because ` "get_{$meta_type}_metadata"` fires prior to checking the objects cache.

As you've been trying to avoid unnecessary database queries while getting post meta data, this acheives that by only processing the queue if a database query will be made.

spacedmonkey reviewed (commented):
One future improvement to the lazy_loading class, we could add a loaded_meta property to the class. This could be used to store a list of id / meta type for meta that has already been loaded. Already loaded meta, would not be allowed to be added to the queue again and would be skipped in the loading.

spacedmonkey commented:
@peterwilsoncc I mocked up an idea that I had to track the already loaded meta ids as part of the metadata lazy loading class. https://github.com/WordPress/wordpress-develop/pull/8399. This would save lot of lookup to cache values. I see this as a general improve to this class and not linked to this change, but it can be bundled, if you see it a blocker 

https://github.com/WordPress/wordpress-develop/pull/8399

peterwilsoncc commented:
@spacedmonkey I don't think it's a blocker, what is a blocker is failing to process the queue for post types when a databse query will be made. The whole point of lazy loaded meta data is to process the queue alongside another query, the custom code for post meta defeats the purpose of this code.

This test, which is currently failing, demonstrates the issue:

```php
/**
 * @ticket 57496
 */
public function test_post_meta_database_queries_process_the_queue() {
	$lazy_loaded_post_ids = self::$post_ids;
	// Remove one post from the queue.
	$unqueued_post_id = array_shift( $lazy_loaded_post_ids );

	new WP_Query(
		array(
			'post_type'           => 'wptests_pt',
			'post__in'            => $lazy_loaded_post_ids,
			'lazy_load_post_meta' => true,
		)
	);

	// Post meta cache should be empty.
	$cache = wp_cache_get_multiple( self::$post_ids, 'post_meta' );
	$cache = array_filter( $cache );
	$this->assertEmpty( $cache, 'Post meta cache is expected to be empty' );

	// Getting post meta for unqueued post should trigger a database query.
	$queries_before = get_num_queries();
	get_post_meta( $unqueued_post_id );
	$queries_after = get_num_queries();
	$this->assertSame( 1, $queries_after - $queries_before, 'Expected a database query getting unqueued post meta' );

	// Ensure getting post meta for queued posts does not trigger a database query.
	$queries_before = get_num_queries();
	array_map( 'get_post_meta', self::$post_ids );
	$queries_after = get_num_queries();
	$this->assertSame( 0, $queries_after - $queries_before, 'Getting lazy loaded post meta is not expected to trigger a database query' );

	// Ensure all the post meta is now in the cache.
	$cache = wp_cache_get_multiple( self::$post_ids, 'post_meta' );
	$cache = array_filter( $cache );
	$this->assertCount( count( self::$post_ids ), $cache, 'Expected all post meta to be in the cache' );
}
```

Removing the custom code for post meta allows the test to pass.

spacedmonkey commented:
> I don't think it's a blocker, what is a blocker is failing to process the queue for post types when a databse query will be made. The whole point of lazy loaded meta data is to process the queue alongside another query, the custom code for post meta defeats the purpose of this code.
> 
> This test, which is currently failing, demonstrates the issue:

@peterwilsoncc 
This PR is not just about reducing database queries by bundling post meta in the queue when post meta is loaded. Let me clarify.

For blog (site), term, and comment meta, WordPress core rarely, if ever, makes use of this data. These types of meta are primarily designed for developers to extend functionality. Because of this, it would be inefficient to load them on every page request, which is why they are lazily loaded.

Post meta, however, is different. It is widely used throughout WordPress core across all post types. Unlike blog, term, or comment meta functions, get_post_meta is called frequently—sometimes over 100 times per page load. As a result, the post meta queue would be constantly cleared, making lazy loading ineffective.

This PR does not aim to bundle post meta queries together. Instead, it creates a queue of post meta IDs that might be needed in a request. Here are two examples:
	1.	Post meta for parent posts of posts within the loop.
	2.	Post meta for posts/pages that are referenced within other posts.

The first time an ID in the queue is requested—such as the first menu item linking to a page—all associated meta will be loaded at once. This allows us to disable post meta priming in areas like menus, where the post meta for linked pages is unlikely to be needed. See [#60199](https://core.trac.wordpress.org/ticket/60199).

By design, your test would fail, as it is not designed to do what you are asking.

peterwilsoncc commented:
> This PR does not aim to bundle post meta queries together. Instead, it creates a queue of post meta IDs that might be needed in a request. Here are two examples: 1. Post meta for parent posts of posts within the loop. 2. Post meta for posts/pages that are referenced within other posts.
> 
> The first time an ID in the queue is requested—such as the first menu item linking to a page—all associated meta will be loaded at once. This allows us to disable post meta priming in areas like menus, where the post meta for linked pages is unlikely to be needed. See [#60199](https://core.trac.wordpress.org/ticket/60199).

I think what you are describing is the the purpose of the existing `update_post_meta_cache` option in `WP_Query`.

The design of lazy loading post meta IS to clear the queue when post meta is requested. Having it behave differently for different types of meta data is simply confusing, and won't result in fewer database queries.

The concern in Core-60199 is a valid concern but the code as written doesn't solve that. Instead of querying post meta in a single query, it increases the chance that post meta will be loaded for each post as a single query.

If loading post meta by single queries is better under some circumstances, then the PR should be targettng that rather than using the existing API.

joemcgill commented on src/wp-includes/class-wp-query.php:
```diff
@@ -797,9 +797,8 @@ public function fill_query_vars( $query_vars ) {
-	 *     @type bool            $lazy_load_post_meta    Whether to lazy-load post meta. Setting to false will
-	 *                                                   disable cache priming for post meta, so that each
-	 *                                                   get_post_meta() call will hit the database.
+	 *     @type bool            $lazy_load_post_meta    Whether to lazy-load post meta. If set to false, post meta priming will
+	 *                                                   fallback to the default behavior of priming post meta.
```
I think the following changes better reflect the current state of the logic and resolves conversations [here](https://github.com/WordPress/wordpress-develop/pull/4216/files/434cb55fca1d6d86d3f230256f02ac9333943992..34855d127635aa49606cbd81293dbad8b2c616e9#r1878491731) and [here](https://github.com/WordPress/wordpress-develop/pull/4216/files/434cb55fca1d6d86d3f230256f02ac9333943992..34855d127635aa49606cbd81293dbad8b2c616e9#r1878494301).

```suggestion
	 *     @type bool            $lazy_load_post_meta    Whether to lazy-load post meta. When set to true, post meta priming will
	 *                                                   be disabled. Default false.
```

joemcgill commented:
Re-reading the current discussion and if I'm understanding correctly, the only question that needs to be resolved is whether queued post meta should be loaded only when another **queued** post meta is called, rather than clearing the queue when **any uncached** post meta is loaded.

The current approach seems to only load queued meta when another queued meta ID is requested.

I think @peterwilsoncc's argument in [this comment](https://github.com/WordPress/wordpress-develop/pull/4216#discussion_r1938610347) is valid. To paraphrase:

The risk of loading unused post meta is just as likely when loading another ID from the queue as it is when loading an ID that was not already in the queue. However, I don't feel strongly about this being a blocker to an initial commit. If we find that only clearing the queue when a lazy loaded ID is called results in more unnecessary DB calls, then we can adjust this before the final release.

swissspidy commented:
I'm probably not the right one to review this...are there perhaps other post meta experts?

NB: there are some merge conflicts now, so a refresh would be good.

swissspidy commented on src/wp-includes/class-wp-query.php:
```diff
@@ -660,6 +660,7 @@ public function fill_query_vars( $query_vars ) {
 	 * @since 5.3.0 Introduced the `$meta_type_key` parameter.
 	 * @since 6.1.0 Introduced the `$update_menu_item_cache` parameter.
 	 * @since 6.2.0 Introduced the `$search_columns` parameter.
+	 * @since 6.8.0 Introduced the `$lazy_load_post_meta` parameter.
```
This needs updating now.

swissspidy commented on src/wp-includes/post.php:
```diff
@@ -2665,6 +2665,21 @@ function get_post_meta( $post_id, $key = '', $single = false ) {
+/**
+ * Queue post meta for lazy-loading.
+ *
+ * @since 6.8.0
```
This needs updating now.

SirLouen commented on src/wp-includes/class-wp-query.php:
```diff
@@ -1291,7 +1294,7 @@ public function parse_tax_query( &$q ) {
 			$q['category__not_in'] = array_map( 'absint', array_unique( (array) $q['category__not_in'] ) );
 			sort( $q['category__not_in'] );
-			$tax_query[]           = array(
+			$tax_query[] = array(
```
This is one that is not applying because it was already fixed

SirLouen commented on src/wp-includes/class-wp-query.php:
```diff
@@ -1360,7 +1363,7 @@ public function parse_tax_query( &$q ) {
 			$q['tag__not_in'] = array_map( 'absint', array_unique( (array) $q['tag__not_in'] ) );
 			sort( $q['tag__not_in'] );
-			$tax_query[]      = array(
+			$tax_query[] = array(
```
This is the second that is not applying

SirLouen commented on src/wp-includes/class-wp-query.php:
```diff
@@ -3744,6 +3770,9 @@ public function the_post() {
 			$post_ids = array_filter( $post_ids );
 			if ( $post_ids ) {
 				_prime_post_caches( $post_ids, $this->query_vars['update_post_term_cache'], $this->query_vars['update_post_meta_cache'] );
+				if ( $this->query_vars['lazy_load_post_meta'] ) {
```
This changeset [[59993](https://core.trac.wordpress.org/changeset/59993)] brought major changes in this section.

SirLouen reviewed (changes requested):
@spacedmonkey I've added you some comments to help refreshing the PR

Please ping me in trac when you have finished refreshing this (and maybe adding the proposed changes by the rest) so I can start some performance testing.
